### PR TITLE
Loosen CPI restrictions and charge compute for ix data len

### PR DESCRIPTION
**Pull request by jstarry**
#### Problem
- When address lookup tables are enabled, the number of accounts that can be passed in a CPI is lower than what can be passed in a transaction instruction.
- Only the accounts passed to a CPI incur compute unit costs based on the length of their data fields. Instruction data size doesn't incur compute unit cost. This isn't really an issue now because CPI size restrictions are strict. But once the restrictions are loosened, there should be a cost.

#### Summary of Changes
- Loosen restrictions on CPI size
- With looser restrictions on CPI size, charge compute units for the size of ix data

Fixes https://github.com/solana-labs/solana/issues/26639
Feature Gate Issue: https://github.com/solana-labs/solana/issues/26641

### CODE REVIEW

File: programs/bpf/c/src/invoke/invoke.c [hunk 1]
```diff
@@ -480,28 +481,29 @@ extern uint64_t entrypoint(const uint8_t *input) {
...
 
     break;
   }
-  case TEST_INSTRUCTION_META_TOO_LARGE: {
-    sol_log("Test instruction meta too large");
-    SolAccountMeta *arguments = sol_calloc(40, sizeof(SolAccountMeta));
-    sol_log_64(0, 0, 0, 0, (uint64_t)arguments);
+  case TEST_MAX_INSTRUCTION_ACCOUNTS_EXCEEDED: {
+    sol_log("Test max instruction accounts exceeded");
+    uint64_t accounts_len = 256;
```
Comment (Lichtso) [inline programs/bpf/c/src/invoke/invoke.c:501]: Magic number, might be using a `const` from a definition somewhere else?

File: programs/bpf/rust/invoke/src/processor.rs [hunk 2]
```diff
@@ -554,21 +554,27 @@ fn process_instruction(
...
+            let default_key = Pubkey::default();
+            let account_metas = vec![(&default_key, false, false); 256];
+            let instruction =
+                create_instruction(*accounts[INVOKED_PROGRAM_INDEX].key, &account_metas, vec![]);
             invoke_signed(&instruction, &[], &[])?;
         }
+        TEST_MAX_ACCOUNT_INFOS_EXCEEDED => {
+            msg!("Test max account infos exceeded");
+            let instruction = create_instruction(*accounts[INVOKED_PROGRAM_INDEX].key, &[], vec![]);
+            let account_infos = vec![accounts[0].clone(); 65];
```
Comment (Lichtso) [inline programs/bpf/rust/invoke/src/processor.rs:575]: Also magic number.

File: programs/bpf_loader/src/syscalls/cpi.rs [hunk 3]
```diff
@@ -733,41 +733,104 @@ where
...
-        .saturating_mul(size_of::<AccountMeta>())
-        .saturating_add(data_len);
-    let max_size = invoke_context.get_compute_budget().max_cpi_instruction_size;
-    if size > max_size {
-        return Err(SyscallError::InstructionTooLarge(size, max_size).into());
+    if invoke_context
+        .feature_set
+        .is_active(&feature_set::loosen_cpi_size_restriction::ID)
+    {
+        let data_len = u64::try_from(data_len).unwrap();
```
Comment (Lichtso) [inline programs/bpf_loader/src/syscalls/cpi.rs:759]: Why the explicit `unwrap()` and not an `as u64`?

File: programs/bpf/c/src/invoke/invoke.c [hunk 4]
```diff
@@ -76,7 +77,7 @@ uint64_t do_nested_invokes(uint64_t num_nested_invokes,
 }
 
 extern uint64_t entrypoint(const uint8_t *input) {
-  sol_log("Invoke C program");
+  sol_log("invoke C program");
```
[the remark this answers is not in the document]
Comment (jstarry) [reply programs/bpf/c/src/invoke/invoke.c:80]: Yup, in order to differentiate the errors produced by this feature, the tests now check which log messages are produced. This change made the C example program logs more consistent with the Rust example program logs.

File: programs/bpf/c/src/invoke/invoke.c [hunk 5]
```diff
@@ -510,6 +512,23 @@ extern uint64_t entrypoint(const uint8_t *input) {
 
     break;
   }
+  case TEST_MAX_ACCOUNT_INFOS_EXCEEDED: {
+    sol_log("Test max account infos exceeded");
+    SolAccountMeta arguments[] = {};
+    uint64_t account_infos_len = 65;
```
[the remark this answers is not in the document]
Comment (jstarry) [reply programs/bpf/c/src/invoke/invoke.c]: Fixed

File: programs/bpf/c/src/invoke/invoke.c (hunk 1 above)
Comment (jstarry) [reply programs/bpf/c/src/invoke/invoke.c]: Fixed, added constants in `cpi.h`

File: programs/bpf_loader/Cargo.toml [hunk 6]
```diff
@@ -25,6 +25,7 @@ thiserror = "1.0"
 [dev-dependencies]
 rand = "0.7.3"
 solana-runtime = { path = "../../runtime", version = "=1.11.4" }
+static_assertions = "1.1.0"
```
Comment (jackcmay) [inline programs/bpf_loader/Cargo.toml:28]: why is this necessary in this pr, can we break it out into a separate pr?

File: programs/bpf_loader/src/syscalls/cpi.rs [hunk 7]
```diff
@@ -733,41 +739,89 @@ where
     Ok(accounts)
 }
 
+#[cfg(test)]
+static_assertions::const_assert_eq!(ACCOUNT_META_SIZE, 34);
```
Comment (jackcmay) [inline programs/bpf_loader/src/syscalls/cpi.rs:743]: Looks like this belongs in a sepeerate pr

File: programs/bpf_loader/src/syscalls/cpi.rs [hunk 8]
```diff
@@ -733,41 +739,89 @@ where
...
 ) -> Result<(), EbpfError<BpfError>> {
-    let size = num_accounts
-        .saturating_mul(size_of::<AccountMeta>())
-        .saturating_add(data_len);
-    let max_size = invoke_context.get_compute_budget().max_cpi_instruction_size;
-    if size > max_size {
-        return Err(SyscallError::InstructionTooLarge(size, max_size).into());
+    if invoke_context
+        .feature_set
+        .is_active(&feature_set::loosen_cpi_size_restriction::ID)
```
Comment (jackcmay) [inline programs/bpf_loader/src/syscalls/cpi.rs:753]: for consistency sake `loosen_cpi_size_restrictions::id()`

File: programs/bpf_loader/src/syscalls/cpi.rs [hunk 9]
```diff
@@ -733,41 +739,89 @@ where
...
+        let max_accounts = MAX_CPI_INSTRUCTION_ACCOUNTS as u64;
+        if num_accounts > max_accounts {
+            return Err(SyscallError::MaxInstructionAccountsExceeded {
+                num_accounts,
+                max_accounts,
+            }
+            .into());
+        }
+
+        invoke_context.get_compute_meter().consume(
```
Comment (jackcmay) [inline programs/bpf_loader/src/syscalls/cpi.rs:775]: this is werid to have here, this function checks sizes, how about moving the consume up with the rest of the consumption stuff.  I'd be fine if you want to move it later during feature cleanup to simplify the featurization now

File: sdk/program/src/lib.rs [hunk 10]
```diff
@@ -603,7 +603,6 @@ pub mod slot_hashes;
 pub mod slot_history;
 pub mod stake;
 pub mod stake_history;
-#[cfg(target_os = "solana")]
```
Comment (jackcmay) [inline sdk/program/src/lib.rs:606]: What was this for and why removed here?

**Review by jackcmay — DISMISSED** [review]
Besides some nits, lgtm

File: programs/bpf_loader/src/syscalls/cpi.rs (hunk 8 above)
Comment (jstarry) [reply programs/bpf_loader/src/syscalls/cpi.rs]: Updated

File: programs/bpf_loader/src/syscalls/cpi.rs (hunk 7 above)
Comment (jstarry) [reply programs/bpf_loader/src/syscalls/cpi.rs]: Removed

File: programs/bpf_loader/Cargo.toml (hunk 6 above)
Comment (jstarry) [reply programs/bpf_loader/Cargo.toml]: Not necessary, removed

**Comment by jstarry** [comment]
@jackcmay @Lichtso do the latest updates look good?

**Comment by jstarry** [comment]
@jackcmay @Lichtso I've merged but still would appreciate one more look on the latest changes that addressed the last round of feedback

**Comment by Lichtso** [comment]
LGTM

File: sdk/bpf/c/inc/sol/cpi.h [hunk 11]
```diff
@@ -11,6 +11,28 @@
...
+static const uint8_t MAX_CPI_INSTRUCTION_ACCOUNTS = 255;
+
+/**
+ * Maximum number of account info structs that can be used in a single CPI
+ * invocation. A limit on account info structs is effectively the same as
+ * limiting the number of unique accounts. 64 was chosen to match the max
+ * number of locked accounts per transaction (MAX_TX_ACCOUNT_LOCKS).
+ */
+static const uint8_t MAX_CPI_ACCOUNT_INFOS = 64;
+
```
Comment (dmakarov) [inline sdk/bpf/c/inc/sol/cpi.h:35]: Please, do not change sdk/bpf/c/inc/sol/*.h files directly. They are generated from sdk/bpf/c/inc/sol/inc/*.inc files.

Comment (jstarry) [reply sdk/bpf/c/inc/sol/cpi.h:35]: My bad, can we add them to the gitignore then? Or at least add a comment that says it's generated?

Comment (dmakarov) [reply sdk/bpf/c/inc/sol/cpi.h:35]: The generated files are also committed to the repository, we don't want to ignore them. It's not ideal to commit the generated files, but the assumption is that the header files are changed very rarely and it wouldn't be too much overhead to maintain them in repository. The generated files are currently regenerated when `cargo-install-all.sh` script runs. There are comments in the generated files, however, these header are placed above the syscall declarations that are generated by the tool that converts .inc files to corresponding .h files.  I understand the current situation is vulnerable and error prone but at the moment I have no better solution because the syscalls hash codes need to be generated.

Comment (jstarry) [reply sdk/bpf/c/inc/sol/cpi.h:35]: Ok, that's fine. Thanks for explaining!
